Implement RBAC in the UI

ryelo opened the pull request:
## What
implement RBAC roles in the UI

## Jira
https://projects.engineering.redhat.com/browse/RHCLOUD-4500

## Screenshots

### No perms, but entitled
![Screen Shot 2020-02-27 at 3 17 12 PM](https://user-images.githubusercontent.com/12520842/75483159-406a2d00-5974-11ea-9831-98754c7a412a.png)


### Read
![Screen Shot 2020-02-27 at 1 59 33 PM](https://user-images.githubusercontent.com/12520842/75476923-6ccc7c00-5969-11ea-8c92-48dbec1d7e25.png)
![Screen Shot 2020-02-27 at 1 59 40 PM](https://user-images.githubusercontent.com/12520842/75476934-6fc76c80-5969-11ea-915d-ff9859dc045b.png)


### Read+ Write
![Screen Shot 2020-02-27 at 1 56 59 PM](https://user-images.githubusercontent.com/12520842/75476765-2c6cfe00-5969-11ea-9042-82b99f1a4ecb.png)
![Screen Shot 2020-02-27 at 1 57 06 PM](https://user-images.githubusercontent.com/12520842/75476767-2e36c180-5969-11ea-9c98-e9a71e5957e1.png)

### Read + Write + Execute
![Screen Shot 2020-02-27 at 1 58 40 PM](https://user-images.githubusercontent.com/12520842/75476836-4d355380-5969-11ea-83e5-6fba1b267630.png)
![Screen Shot 2020-02-27 at 1 58 48 PM](https://user-images.githubusercontent.com/12520842/75476844-4f97ad80-5969-11ea-928e-953d8dc9a151.png)

ryelo commented:
Make sure to turn on the fifi flag to see the execute button 😄

jharting commented on src/App.js:
```diff
@@ -8,11 +8,40 @@ import './App.scss';
+                );
+            }
+        );
+        window.insights.chrome.auth.getUser().then((user) => user.identity.user.is_org_admin && this.setState({ isOrgAdmin: true }));
```
This calback may complete at different time than the permission one. I.e. an org admin without permissions would see "you cannot use this" blip first and then it would work. Instead I think we should wait for both these calls to finish before dismissing the loading screen

ryelo commented:
@jharting 
* I reworked some of the consumers to use the hook implementation of context.
* removed the "remove action" from the skeleton. I'd say we remove it for now, but if it's ever decided to just disable buttons instead of hiding them, we can add it back and just keep it disabled
* I removed the org admin check
* removed the checkboxes if you don't have write perms on the details page

jharting commented on src/App.js:
```diff
@@ -8,11 +8,38 @@ import './App.scss';
     componentDidMount () {
         insights.chrome.init();
         insights.chrome.identifyApp('remediations');
+        window.insights.chrome.getUserPermissions().then(
```
This fetches all permissions which is wasteful. It should be enough to pass "remediations" to the function

jharting replied:
However, even when I change the code to pass the app name it still does not seem to be making the correct REST request :-/

ryelo replied:
This is currently calling the chrome function. I don't believe we have a way to filter out a certain app at this time.

jharting replied:
@ryelo isn't it what this first param does? https://github.com/RedHatInsights/insights-chrome/blob/master/src/js/entry.js#L125

jharting replied:
Or rather supposed to do? Like I said when testing it did not seem to work

karelhala replied:
It is exactly what @jharting is refering to, it is strange though because I tested it and it does exactly what was expected. It queries the RBAC server with applications parameter in query URL

ryelo replied:
This is working for me! I'll update accordingly

jharting commented on src/App.js:
```diff
@@ -8,11 +8,38 @@ import './App.scss';
         insights.chrome.init();
         insights.chrome.identifyApp('remediations');
+        window.insights.chrome.getUserPermissions().then(
+            allPermissions => {
```
Should we delay loading of data (i.e. keep a spinner on) until after this finishes? Otherwise a data fetch request will cause 403 and an error toast in addition to the DeniedState

ryelo commented on src/App.js:
```diff
@@ -8,11 +8,38 @@ import './App.scss';
+            allPermissions => {
+                const permissionList = allPermissions.map(permissions => permissions.permission);
+                this.handlePermissionUpdate(
+                    permissionList.includes('remediations:remediation:read'),
```
Are you saying that if a user has any `remediations:foo:bar` permission, they also get a "read" permission?

jharting replied:
@ryelo Not at all. Each permission has three segments separated by colons i.e. application:resource_type:operation. Each of these may either have a constant value or asterisk may be used as a wildcard. If an orgadmin gives somebody a `remediations:*:*` permission it means they can do whatever in Remediations. The orgadmin may decide to assign explicit constant permissions (e.g. `remediations:remediation:read` and `remediations:remediation:write`), use a wildcard (e.g. `remediations:remediation:*`) or use a combination. Therefore, we should enable "write" controls not only in cases where we discover user having the constant write permission (`remediations:remediation:write`) but also look for wildcard ones that match.

See https://docs.google.com/document/d/1x1kbj7YvEz4IOWEp2-wis2OlvUCcVwkcc4aTPsfRIAU/edit# for more details.

jharting commented on src/routes/Home.js:
```diff
@@ -69,24 +72,32 @@ class Home extends Component {
-            </React.Fragment>
+            <PermissionContext.Consumer>
+                { value =>
+                    value.permissions.read === false &&
```
This should only depend on the read permission - i.e. if the user does not have read permission they won't see anything (as the API won't return data without the read permission)

jharting commented on src/routes/RemediationDetails.js:
```diff
@@ -143,11 +145,15 @@ class RemediationDetails extends Component {
-                                            remediationId={ remediation.id }>
-                                        </ExecutePlaybookButton>
-                                    </SplitItem>
+                                    <PermissionContext.Consumer>
```
Can we add the `<DeniedState/>` logic all routes (meaning here)? I.e. if the read permission is taken away from me `/remediations` gets me to `<DeniedState/>` but so should `/remediations/12345` (e.g. an old bookmark)
